[GTK][WPE] Add a content type property to TextureMapperPlatformLayerProxy

magomez opened the pull request:
#### ba0578f34b1d565deb05a1db81902a9a0d925080
<pre>
[GTK][WPE] Add a content type property to TextureMapperPlatformLayerProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=272112">https://bugs.webkit.org/show_bug.cgi?id=272112</a>

Reviewed by Carlos Garcia Campos.

Add a property to TextureMapperPlatformLayerProxy to indicate the type of
frames that are passing through it. The type can be WebGL, Video,
OffscreenCanvas or HolePuch.

We don&apos;t want to have proxies without a defined content type, so this patch
removes the method to create a Nicosia::ContentLayer that doesn&apos;t receive
a proxy and creates one whose content type is not known. This way, the code
that creates the Nicosia::ContentLayer has to specifically create a proxy
with a defined content type for it.

This change removes the need of the of the disableBufferInvalidation parameter
in the TextureMapperPlatformLayerProxyGL constructor, as this is only used for
the HolePunch case. Now that the proxy knows when it&apos;s handling HolePunch
buffers, the previous condition can be reworked.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::MediaPlayerPrivateHolePunch):
* Source/WebCore/platform/graphics/nicosia/NicosiaContentLayer.cpp:
* Source/WebCore/platform/graphics/nicosia/NicosiaContentLayer.h:
* Source/WebCore/platform/graphics/nicosia/NicosiaGCGLANGLELayer.cpp:
(Nicosia::GCGLANGLELayer::GCGLANGLELayer):
* Source/WebCore/platform/graphics/nicosia/NicosiaImageBufferPipe.cpp:
(Nicosia::NicosiaImageBufferPipeSource::NicosiaImageBufferPipeSource):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.cpp:
(WebCore::TextureMapperPlatformLayerProxy::TextureMapperPlatformLayerProxy):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
(WebCore::TextureMapperPlatformLayerProxy::contentType const):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
(WebCore::TextureMapperPlatformLayerProxyDMABuf::TextureMapperPlatformLayerProxyDMABuf):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::TextureMapperPlatformLayerProxyGL):
(WebCore::TextureMapperPlatformLayerProxyGL::invalidate):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebKit/WebProcess/GPU/graphics/gbm/RemoteGraphicsContextGLProxyGBM.cpp:
(WebKit::NicosiaDisplayDelegate::NicosiaDisplayDelegate):

Canonical link: <a href="https://commits.webkit.org/277110@main">https://commits.webkit.org/277110@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e69cfbfad47793baa760a0eee7c24f341a58214

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46724 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49345 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49401 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42771 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49031 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23347 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47304 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19354 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/20228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41377 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4770 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41744 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51278 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21733 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18093 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23021 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44308 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10322 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
```diff
@@ -53,6 +61,9 @@ class TextureMapperPlatformLayerProxy : public ThreadSafeRefCounted<TextureMappe
     Lock& lock() WTF_RETURNS_LOCK(m_lock) { return m_lock; }
     bool isActive();
 
+    ContentType contentType() { return m_contentType; }
```
This may be `const`:

```suggestion
    ContentType contentType() const { return m_contentType; }
```

aperezdc reviewed (commented):
LGTM, as this is changing from a boolean flag to an `enum` (which is almost always a good thing), making the code clearer. I left only a minor comment.

I would approve this, as there are no functional changes otherwise, but I'll defer the final rubber-stamp to a multimedia-savvy reviewer.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
```diff
@@ -39,12 +39,20 @@ class TextureMapperPlatformLayerBuffer;
     };
 
-    TextureMapperPlatformLayerProxy();
+    TextureMapperPlatformLayerProxy(ContentType);
```
explicit

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.h:
```diff
@@ -45,7 +45,7 @@ class TextureMapper;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    TextureMapperPlatformLayerProxyDMABuf();
+    TextureMapperPlatformLayerProxyDMABuf(ContentType);
```
explicit

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
```diff
@@ -44,7 +44,7 @@ namespace WebCore {
     WTF_MAKE_FAST_ALLOCATED();
 public:
-    TextureMapperPlatformLayerProxyGL(bool disableBufferInvalidation = false);
+    TextureMapperPlatformLayerProxyGL(ContentType);
```
explicit

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaContentLayer.cpp:
```diff
@@ -35,7 +35,7 @@ namespace Nicosia {
 Ref<ContentLayer> ContentLayer::create(Client& client)
 {
-    return adoptRef(*new ContentLayer(client, adoptRef(*new WebCore::TextureMapperPlatformLayerProxyGL)));
+    return adoptRef(*new ContentLayer(client, adoptRef(*new WebCore::TextureMapperPlatformLayerProxyGL(WebCore::TextureMapperPlatformLayerProxy::ContentType::Undefined))));
```
What is this for? It's the only case where undefined is used (apart from initial value), can we "define" it somehow?

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaImageBufferPipe.cpp:
```diff
@@ -48,6 +48,7 @@ using namespace WebCore;
 NicosiaImageBufferPipeSource::NicosiaImageBufferPipeSource()
 {
     m_nicosiaLayer = Nicosia::ContentLayer::create(*this);
+    m_nicosiaLayer->proxy().setContentType(TextureMapperPlatformLayerProxy::ContentType::OffscreenCanvas);
```
Ah, it's OffscreenCanvas but set after construction, right?. It's unfortunate we can't set it on construction. Maybe it would be less confusing to make `TextureMapperPlatformLayerProxyGL` have a default value to undefined and not give any argument in `Nicosia::ContentLayer::create()`.

carlosgcampos replied:
Or can we use the `Nicosia::ContentLayer::create()` that receives the proxy instead? And we don't need the setter, making it a construct only property.

magomez replied:
> Or can we use the `Nicosia::ContentLayer::create()` that receives the proxy instead? And we don't need the setter, making it a construct only property.

I think that's a good idea, but I think then we should remove the `Nicosia::ContentLayer::create()` version that doesn't receive a proxy to ensure that the proxies need to be specifically created with a concrete contentType.

carlosgcampos replied:
Let's do it if it's unused.

magomez committed:
[GTK][WPE] Add a content type property to TextureMapperPlatformLayerProxy
https://bugs.webkit.org/show_bug.cgi?id=272112

Reviewed by Carlos Garcia Campos.

Add a property to TextureMapperPlatformLayerProxy to indicate the type of
frames that are passing through it. The type can be WebGL, Video,
OffscreenCanvas or HolePuch.

We don't want to have proxies without a defined content type, so this patch
removes the method to create a Nicosia::ContentLayer that doesn't receive
a proxy and creates one whose content type is not known. This way, the code
that creates the Nicosia::ContentLayer has to specifically create a proxy
with a defined content type for it.

This change removes the need of the of the disableBufferInvalidation parameter
in the TextureMapperPlatformLayerProxyGL constructor, as this is only used for
the HolePunch case. Now that the proxy knows when it's handling HolePunch
buffers, the previous condition can be reworked.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::MediaPlayerPrivateGStreamer):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::MediaPlayerPrivateHolePunch):
* Source/WebCore/platform/graphics/nicosia/NicosiaContentLayer.cpp:
* Source/WebCore/platform/graphics/nicosia/NicosiaContentLayer.h:
* Source/WebCore/platform/graphics/nicosia/NicosiaGCGLANGLELayer.cpp:
(Nicosia::GCGLANGLELayer::GCGLANGLELayer):
* Source/WebCore/platform/graphics/nicosia/NicosiaImageBufferPipe.cpp:
(Nicosia::NicosiaImageBufferPipeSource::NicosiaImageBufferPipeSource):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.cpp:
(WebCore::TextureMapperPlatformLayerProxy::TextureMapperPlatformLayerProxy):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxy.h:
(WebCore::TextureMapperPlatformLayerProxy::contentType const):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
(WebCore::TextureMapperPlatformLayerProxyDMABuf::TextureMapperPlatformLayerProxyDMABuf):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::TextureMapperPlatformLayerProxyGL):
(WebCore::TextureMapperPlatformLayerProxyGL::invalidate):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebKit/WebProcess/GPU/graphics/gbm/RemoteGraphicsContextGLProxyGBM.cpp:
(WebKit::NicosiaDisplayDelegate::NicosiaDisplayDelegate):

Canonical link: https://commits.webkit.org/277110@main
